Feature/json editor

jnnwnk opened the pull request:
To make it easier to edit the questionnaire in JSON format, the use of a JSON editor will be evaluated.

Currently the following editor is in use: https://github.com/josdejong/jsoneditor

jnnwnk commented:
After adding the dependency, `npm audit` shows some vulnerabilities. We should have a look how to fix those, if we decide to use this editor or have a look for another editor if it's not possible to fix.


```
$ npm audit

                       === npm audit security report ===                        


                                 Manual Review                                  
             Some vulnerabilities require your attention to resolve             
                                                                                
          Visit https://go.npm.me/audit-guide for additional guidance           


  Low             Insecure Entropy Source - Math.random()                       

  Package         node-uuid                                                     

  Patched in      >=1.4.4                                                       

  Dependency of   jsoneditor                                                    

  Path            jsoneditor > ace-builds > ace > node-uuid                     

  More info       https://npmjs.com/advisories/93                               


  Low             Regular Expression Denial of Service                          

  Package         debug                                                         

  Patched in      >= 2.6.9 < 3.0.0 || >= 3.1.0                                  

  Dependency of   jsoneditor                                                    

  Path            jsoneditor > ace-builds > ace > debug                         

  More info       https://npmjs.com/advisories/534                              


  High            Denial of Service                                             

  Package         https-proxy-agent                                             

  Patched in      >=2.2.0                                                       

  Dependency of   jsoneditor                                                    

  Path            jsoneditor > ace-builds > ace > https-proxy-agent             

  More info       https://npmjs.com/advisories/593                              


  High            Machine-In-The-Middle                                         

  Package         https-proxy-agent                                             

  Patched in      >=2.2.3                                                       

  Dependency of   jsoneditor                                                    

  Path            jsoneditor > ace-builds > ace > https-proxy-agent             

  More info       https://npmjs.com/advisories/1184                             


  High            Arbitrary File Overwrite                                      

  Package         tar                                                           

  Patched in      >=2.2.2 <3.0.0 || >=4.4.2                                     

  Dependency of   jsoneditor                                                    

  Path            jsoneditor > ace-builds > ace > tar                           

  More info       https://npmjs.com/advisories/803                              


  Low             Prototype Pollution                                           

  Package         minimist                                                      

  Patched in      >=0.2.1 <1.0.0 || >=1.2.3                                     

  Dependency of   jsoneditor                                                    

  Path            jsoneditor > ace-builds > ace > minimist                      

  More info       https://npmjs.com/advisories/1179                             

found 6 vulnerabilities (3 low, 3 high) in 932001 scanned packages
  6 vulnerabilities require manual review. See the full report for details.
```

jnnwnk commented:
Another suggestion from @ChristophWersal which should be evaluated (taken from Slack):

> Mit dem Feature "Transform JSON using JMESPath queries." kann man evtl unseren eigenen JsonLogic-Parser von Max einhängen... (nicht genauer geschaut ob das geht, nur ein Gedanke zum Testen)

DanielHabenicht commented:
Did you had a look on other editors as well? 
I have found this one https://github.com/json-editor/json-editor 

(Just an uneducated guess)

ejoebstl commented:
I think the vulnerabilities are not an issue, because all data that is loaded or displayed comes from our backend and can be considered safe.

Probably we can simply take the editor which looks best/is easiest to work with (and optimally supports validation).

From my side: Please merge. :)

jnnwnk commented:
> Did you had a look on other editors as well?
> I have found this one https://github.com/json-editor/json-editor
> 
> (Just an uneducated guess)

Not yet. I'll have a closer look tomorrow or at the weekend. The current one is also only another first uneducated guess. 😉

jnnwnk commented:
Nachdem ich noch etwas weiter recherchiert habe, habe ich nun mal den folgenden Editor eingebunden: https://react-jsonschema-form.readthedocs.io/en/latest/

Die Doku ist relativ ausführlich und Material UI wird auch direkt mit unterstützt.

Schaut euch gerne den aktuellen Stand schon mal an. Es ist noch absolut ausbaufähig, aber für einen ersten Eindruck sollte es funktionieren (auch wenn die Tests gerade nicht durchlaufen).

Bei den Auswahlboxen (z.B. Fragetyp) war das Problem, dass dort erstmal nur "Option 1", "Option 2" usw. angezeigt wurde. Um dort etwas Sprechendes anzuzeigen, musste ich die Generierung des Schemas verändern (siehe https://github.com/CovOpen/CovQuestions/pull/10/commits/c47497b1d0c5ee8834a8e509e2a52a03b3542fa3 ). Ich habe dafür quasi erstmal eine Kopie von dem Projekt gemacht und mit hier ins Repo gepackt, damit jeder auf die Generierung zugreifen und sich das anschauen kann.


Weitere Schritte / Ideen wären:
- Todo: Tests fixen!!
- Todo: Styling anpassen, so dass es übersichtlich(er) wird
- Todo: Weitere Untersuchungen bzgl. Generierung des JSON Schemas (ggf. echter Fork von dem Projekt oder dort ein Pull-Request)
- Todo: Kommentare im `schema.ts` und `logic.ts` erweitern, so dass für alle notwendigen Typen sprechende Namen generiert werden
- Idee: Man kann die Felder, die für das Schema erzeugt werden customizen (https://react-jsonschema-form.readthedocs.io/en/latest/form-customization/). Hier könnte man den Logic Parser von @max-simon als Grundlage nehmen und ein einfaches Textfeld anzeigen, wo dann die Bedingungen für `enableWhen` formuliert werden können. Das würde es vermutlich einfacher machen als dort die JSON Objekte von Hand zu schreiben
- Idee: Evtl. Umschaltung zwischen JSON Objekt in einem Textfeld (wie vorher) und dem Formular erlauben (z.B. über zwei Tabs o.ä.). Das wäre für Testzwecke für uns vermutlich hilfreich und andere Benutzer haben später die Wahl wie sie einen Fragebogen erstellen wollen.
- Idee: Für die Übersicht wäre es bestimmt hilfreich, wenn man einzelne Fragen / Bereiche über ein `Expansion Panel` (https://material-ui.com/components/expansion-panels/) minimieren könnte. Da weiß ich allerdings nicht, ob sich das einfach (vielleicht gibt es ja auch Konfigurationen o.ä. dafür) in das Formular einbauen lässt.

DanielHabenicht commented:
Sind die 300 Dateien des json-schema-generator absicht?

jnnwnk commented:
> Sind die 300 Dateien des json-schema-generator absicht?

Jein, die waren drin, weil ich Änderungen an dem Generator gemacht hatte. Habe ich jetzt aber wieder raus genommen, weil wir ja besprochen hatten, dass wir sowieso auf Grundlage des JSON Schemas arbeiten werden.

jnnwnk commented:
@ejoebstl hatte gestern nach dem Meeting noch vorgeschlagen die Fragen usw. in Tabs aufzuteilen, um die Übersicht zu erhöhen. Tabs haben sich dann als wenig praktikabel erwiesen (aufgrund der Länge der Fragen). Als Alternative dazu habe ich heute eine Art Menü-Struktur implementiert. Ich glaube, die funktioniert ganz gut. Gerne Meinungen dazu.
Aus dem aktuellen Stand würden sich folgende weitere Todos ergeben:
- [ ] Felder für Schema customizen (https://react-jsonschema-form.readthedocs.io/en/latest/form-customization/), Textfeld für `enableWhen` definieren in dem eine Bedingung z.B. nach Javascript oder Python-Syntax formuliert werden kann. Ggf. auf Basis des Logic Parsers von Max
- [ ] Developer Mode implementieren (einfacher JSON Editor), wie das alte Textfeld mit Zusatzfunktionalitäten
- [ ] Hinzufügen von Fragen, Variablen, Ergebnissen implementieren
- [ ] Entfernen von Fragen, Variablen, Ergebnissen implementieren
- [ ] Verschieben von Fragen, Variablen, Ergebnissen implementieren

Feedback ist gern gesehen! 🙂

ejoebstl commented:
That's awesome. Really cool ideas in there. Should we auto-use the questionnaire when changed, so that the changes are visible immediately?

Moving questions up/down could be either implemented using drag drop, or by a two buttons. Whatever is easiest.

ejoebstl commented:
I've cherry picked the Covapp questionaire form the other branch, so we have a larger example to test with.

ChristophWersal commented:
> That's awesome. Really cool ideas in there. Should we auto-use the questionnaire when changed, so that the changes are visible immediately?

Do you mean on every key press in the editor? Or when exactly? We explicitly added this, since we did not have JsonValidation in the textfield before. I guess that we could remove it now and make it immediate again, if the json is valid. The only issue: we would then also see the Json-Invalid message during typing thw whole time, if the schema is currently invalid. Maybe we could use onBlur() instead of onChange() to update the left side automatically?

Also: you only always see the first question, since the questionnaire execution is reset at every change of the json. What we could consider: In "editing mode", we show all questions below each other on the left side. This might give a better overview for the editor. We would then need some button "Test Questionnaire" to actually run it in an "execution mode". (Just a thought...)

jnnwnk commented:
> Maybe we could use onBlur() instead of onChange() to update the left side automatically?
I do not know if onBlur() is a good moment to update the questionnaire. If a user has changed questions and then wants to answer a question on the left side, this would be the moment when the onBlur() event is triggered. This would be a bad moment for the update. Maybe we could have a look at a delayed event handler for typing like this https://stackoverflow.com/a/1909508

We could also add some kind of settings where the user could turn the auto update on/off.

I personally like the questionnaire execution just next to the definition. But maybe we could also talk to Alexander about those questions. Maybe he can give us a few hints on how to make the user interface practical?

jnnwnk commented:
> Moving questions up/down could be either implemented using drag drop, or by a two buttons. Whatever is easiest.

I just added buttons for moving questions, result categories and variables. Drag and Drop might be next version if necessary. I actually don't know if we really need the moving for all types (questions, result categories, variables) or only for questions?

Removing elements is also implemented.

DanielHabenicht commented:
is this relatively stable so we can merge? I Would like to try out the auto-generated Typescript classes :)
